dynawo/sources/Solvers/Common/DYNSolverImpl.cpp

```diff
@@ -213,30 +211,44 @@ Solver::Impl::solve(double tAim, double &tNxt, std::vector<double> &yNxt, std::v
 }
 
 bool
-Solver::Impl::evalZMode(vector<state_g> &G0, vector<state_g> &G1, const double & time, bool& discreteVariableChangeFound) {
+Solver::Impl::evalZMode(vector<state_g> &G0, vector<state_g> &G1, const double & time) {
   Timer timer("SolverIMPL::evalZMode");
   bool zChange = false;
   bool modeChange = false;
+  bool stableRoot = true;
   bool change = false;
   model_->setModeChangeType(NO_MODE);
 
   for (int i = 0; i < 10; ++i) {
     // evalZ
     model_->evalZ(time, vYy_, vYp_, vYz_);
     zChange = model_->zChange();
-    discreteVariableChangeFound |= zChange;
 
     // evalMode
     model_->evalMode(time, vYy_, vYp_, vYz_);
     modeChange = model_->modeChange();
 
     // evaluate G and compare with previous values
-    bool stableRoot = detectUnstableRoot(G0, G1, time);
+    stableRoot = detectUnstableRoot(G0, G1, time);
 
-    if (!zChange && !modeChange && stableRoot)
-      return change;
-    else
+    if (zChange && modeChange) {
+      state_ = ModeAndZChange;
```

Comment:
We might also use a bitmask here to avoid this complex logic

Reply:
New issue #382 opened